### ISSUE TITLE
Yield better results on unequal lengths of result object lists

### ISSUE THREAD

**Issue by yash-nisar** [coala/coala#4482]
We need to take care of cases where the lengths of `observed_result` and `expected_result` differ. We would like inputs from fellow coalians because there are many many cases to keep in mind while framing this snippet.

status/needs design
difficulty/high
area/tests

**Comment by yash-nisar** [comment]
We've come up with a basic idea where we match the first common element in both the lists and compare 1:1 each element until one of the list is finished. Every remaining element is considered an extra element and everything before the first common match is considered an prepended extra result.

**Comment by manankalra** [comment]
- `expected_result`: hardcoded in the test case, `observed_result`: the result that the bear under test yields - Right?
- If yes, in what cases will their length differ?
- If `check_results` asserts even if we provide a sliced-up list of the actual results that were obtained, won't this be misused?

**Comment by Makman2** [comment]
> If yes, in what cases will their length differ?

Yes, if our expectation doesn't match the observation ;) I mean it's testing :)

> If check_results asserts even if we provide a sliced-up list of the actual results that were obtained, won't this be misused?

What do you mean with "misused"? And if you provide a sliced-up list, the test would compare it to your slice, right?

**Comment by manankalra** [comment]
By "misused", this is what I meant:
- Considering https://github.com/coala/coala/issues/4482#issuecomment-316396733, providing a sliced-up list will enable the developer to use as less number of expected results as possible in order to test a bear.

**Comment by Makman2** [comment]
I don't understand that.
